share/eds/ods.go

```diff
@@ -0,0 +1,81 @@
+package eds
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+
+	cbor "github.com/ipfs/go-ipld-cbor"
+	"github.com/ipld/go-car"
+	"github.com/ipld/go-car/util"
+)
+
+var errNilReader = errors.New("odsReader: can't create ODSReadr over nil reader")
+
+type odsReaderBuffered struct {
+	reader                 *bufio.Reader
+	current, odsSquareSize int
+	buf                    *bytes.Buffer
+	err                    error
+}
+
+// ODSReader reads data from io.ReadCloser and limits the reader to the CAR header and first quadrant (ODS)
+func ODSReader(r io.ReadCloser) io.Reader {
+	if r == nil {
+		return &odsReaderBuffered{err: errNilReader}
+	}
+
+	odsR := &odsReaderBuffered{
+		reader: bufio.NewReader(r),
+		buf:    new(bytes.Buffer),
+	}
+
+	// read full header to determine amount of shares needed
+	hb, err := util.LdRead(odsR.reader)
+	if err != nil {
+		odsR.err = fmt.Errorf("read header: %v", err)
+		return odsR
+	}
+
+	var ch car.CarHeader
+	if err := cbor.DecodeInto(hb, &ch); err != nil {
+		odsR.err = fmt.Errorf("invalid header: %v", err)
+		return odsR
+	}
+
+	odsWidth := len(ch.Roots) / 4
```

Comment:
we should make `4` a const. since it is also used in testing.

I assume the nested for loops that are iterating 0 to 4 are the same 4, so they should be linked as to not break testing in the future if this value changes

Reply:
They are different actually. 
 - 4 in mentioned code is based on block architecture.  Shares are rsmt2d encoded so amount of resulting shares is **x4** to amount of original shares. Not sure it needs a const, since it is just physical property of sqaure.
 - 4 in tests is based on squareSize used in `randomEDS` to generate test data. `randomEDS` could be change to take the size as and input param, but I think it is a matter of the original [PR](https://github.com/celestiaorg/celestia-node/pull/1232)